qiskit/ignis/experiments/calibration/instruction_data/gate_and_schedules_definition.py

```diff
@@ -55,243 +78,196 @@ def from_backend(cls, backend: BaseBackend) -> 'InstructionsDefinition':
+        gate = Gate(name=gate_name, num_qubits=len(qubits), params=list(schedule.parameters))
         circ = QuantumCircuit(self._n_qubits)  # Probably a better way of doing this
         circ.append(gate, qubits)
         circ.add_calibration(gate, qubits, schedule, params=schedule.parameters)
 
         return circ
 
-    def get_schedule(self, name: str, qubits: Tuple,
-                     free_parameter_names: Optional[List[str]] = None) -> Schedule:
+    def add_gate_schedule(self,
+                          gate_name: str,
+                          qubits: Union[int, Tuple[int]],
+                          signature: List[str],
+                          schedule: Schedule):
+        """"""
+        # TODO implement this
+        raise NotImplementedError
+
+    def get_gate_schedule(self,
+                          gate_name: Optional[str] = None,
+                          qubits: Optional[Union[int, Tuple[int]]] = None,
+                          gate_id: Optional[str] = None,
+                          free_parameter_names: Optional[List[str]] = None) -> Schedule:
         """
-        Retrieves a schedule from the instructions dictionary.
+        Retrieves a schedule from the instructions data frame.
+
+        User can specify (gate_name, qubits) or unique schedule id to retrieve target schedule.
 
         Args:
-            name: Name of the schedule to retrive.
+            gate_name: Name of the schedule to retrive.
             qubits: qubits for which to get the schedule.
+            gate_id: Unique id of target schedule.
             free_parameter_names: List of parameter names that will be left unassigned.
                 The parameter assignment is done by querying the pulse parameter table.
 
         Returns:
-            schedule for the given (name, qubits) key.
-        """
-        schedule = self._instructions[(name, qubits)]
-
-        if not isinstance(schedule, Schedule):
-            schedule = self._get_composite_instruction(schedule)
-
-        schedule = copy.deepcopy(schedule)
-
-        if not free_parameter_names:
-            free_parameter_names = []
-
-        binding_dict = {}
-        for param in schedule.parameters:
-            if param.name not in free_parameter_names:
-                binding_dict[param] = self._get_parameter_value(param)
-
-        schedule.assign_parameters(binding_dict)
-
-        return schedule
-
-    def _get_parameter_value(self, parameter: ParameterExpression) -> Union[float, int, complex]:
+            schedule for the given input.
         """
-        Helper method.
+        matched_entries = self._find_entries(gate_name, qubits, gate_id)
 
-        TODO This could be simplified by better integrating with PulseTable
-        """
-        inst_name, ch, pulse_type, name = parameter.name.split('.')
-        qubits = self._channel_map[ch]
+        if len(matched_entries) > 1:
+            raise Exception('Multiple entries are found. Database may be broken.')
```

Comment:
This check is good but are there ways to prevent the DB from ever being broken? E.g. in SQL I would think of imposing unique constraints.

Reply:
To solve this we need to define the response of database when the same gate is registered twice (overwrite/create another entry). Overwriting may break parameter table, but creating another entry doesn't update the reference in another schedule.